Quickstart doc for using Thema in Go (#75)
grafana/thema: docs/go-quickstart.md

package ship

import "github.com/grafana/thema"

thema.#Lineage
name: "ship"
seqs: [
	{
		schemas: [
			// v0.0
			{
				firstField: string
                secondField: int
			},
		]
	},
]
```

## Generating Go Types

Once we have a schema defined, We can use it to generate a type in Go. Use the below command to generate a Go type

```
thema lineage gen gotypes -l ship.cue > go_type.go 
```

This should generate a go file has a type `ship` defined . The generated file looks like below

```go
// This file is autogenerated. DO NOT EDIT.
//
// Generated by "thema lineage gen" from lineage defined in ship.cue

package ship

// Ship defines model for ship.
type Ship struct {
	FirstField  string `json:"firstField"`
	SecondField int    `json:"secondField"`
}
```

Cue types are more expressive as compared to Go types, To use the rich information from CUE in GO programs, Users can use Bindings. Using Bindings users get access to the thema Lineage defined in `ship.cue` and can start validating their data against any of the schema.

## Generating Go Bindings

Use the below command to generate a Go binding from `ship.cue`

```
thema lineage gen gobindings -l ship.cue > go_bindings.go
```

> The genearted Go binding will look like below

Author: Yup agreed!

func Lineage(rt *thema.Runtime, opts ...thema.BindOption) (thema.Lineage, error) {
	// Load a build instance from the embedded fs
	inst, err := load.InstancesWithThema(themaFS, path.Dir("ship.cue"))
	if err != nil {
		// Errors at this point indicate a problem with basic loading of .cue file bytes,
		// which typically means the code generator was misconfigured and a path input
		// is incorrect.
		return nil, err
	}

	raw := rt.Context().BuildInstance(inst)

	// Errors here indicate that:
	//   - The parsed path does not exist in the loaded CUE file (["github.com/grafana/thema/errors".ErrValueNotExist])
	//   - The value at the parsed path exists, but does not appear to be a Thema
	//     lineage (["github.com/grafana/thema/errors".ErrValueNotALineage])
	//   - The value at the parsed path exists and is a lineage (["github.com/grafana/thema/errors".ErrInvalidLineage]),
	//     but is invalid due to the violation of some general Thema invariant -
	//     for example, declared schemas don't follow backwards compatibility rules,
	//     lenses are incomplete.
	return thema.BindLineage(raw, rt)
}

// type guards

var _ thema.LineageFactory = Lineage
```

> This sets us up where we have a single Go function, Which we can call and it will load up the `ship.cue` from disk and return a thema Lineage

Reviewer: ```suggestion
We now have a single Go function our program can call, and it will load up the `ship.cue` from disk and return a [`thema.Lineage`](https://pkg.go.dev/github.com/grafana/thema#Lineage)
```

We can then use the Lineage function to validate our data against the schema of choice

## Validating data

We can now define a function which takes the Lineage returned from the `Lineage` function, Select the schema which we want to validate our data against, and then validate the data

```go
func validateInput() {
	lin, _ := Lineage(thema.NewRuntime(cuecontext.New()))
	sch, _ := lin.Schema(thema.SV(0,0))  //thema.SV(0,0) here represents first schema of first sequence
	sch.Validate(...)

> }

Reviewer: i think this section could be stronger if it talked about validation from both the CLI and Go, to illustrate the symmetry between the approaches.

Using the different schema fields i suggested in the other comment, you could add a file to this PR containing the following JSON:

```json
{
    "name": "Theseus",
    "masts": 9,
}
```

Then ask the reader to 
```bash
curl $URL > invalid.json
thema data validate -l ship.cue invalid.json
```

Then, suggest writing a go test file that does the above here in an actual go test. That way they can easily execute and see the results. A [byte slice embed](https://blog.jetbrains.com/go/2021/06/09/how-to-use-go-embed-in-go-1-16/) could save a few lines in the test file by reusing the downloaded file.

Reviewer: Also, an easy way to get the desired `cue.Value` is to use a [JSON Endec](https://pkg.go.dev/github.com/grafana/thema/vmux#NewJSONEndec)

Author: Oo yup, will add this, definitely gives a better understanding

```

We can pass a value of type `cue.Value` to `sch.Validate()` to validate the data against the first schema of the first sequence. You can also use the method `ValidateAny()` to validate the input data against any schema in our Lineage.

## Wrap up

> This tutorial demostarted how you can create an empty Lineage using the thema CLI, define a simple schema and use it to generate Go types and bindings, and use the bindings to validate data against a selected schema in the Lineage.  

Reviewer: ```suggestion
This tutorial demonstrated how you can create an empty Lineage using the thema CLI, define a simple schema and use it to generate Go types and bindings, and use the bindings to validate data against a selected schema in the Lineage. 
```

# Quickstart: Using Thema in Go

The tutorial demonstrates how a user can create an empty Lineage using the thema CLI, define a simple schema and use it to generate Go types and bindings, and use the bindings to validate data against a selected schema in the Lineage.

## Defining a schema
The most fundamental task in thema is writing a lineage. We can use the Thema CLI to generate an empty lineage using the below command:

```
thema lineage init empty --name ship > ship.cue 
```

This command will genearte an empty lineage named `ship` and store the output in a file named `ship.cue`. The genearted lineage looks like below

```cue
package ship

import "github.com/grafana/thema"

thema.#Lineage
name: "ship"
seqs: [
	{
		schemas: [
			// v0.0
			{
				// TODO (delete me - first schema goes here!)
			},
		]
	},
]
```

> Let's define a simple schema as an object containing two fields named `firstField` of type `string` and `secondField` of type `int`,  

Author: Yup, Used the ship example above

```cue
package ship

import "github.com/grafana/thema"

thema.#Lineage
name: "ship"
seqs: [
	{
		schemas: [
			// v0.0
			{

> 				firstField: string

Reviewer: Let's definitely include some field comments right from the start, so that the user immediately has the impression that documenting their fields is important.

Author: Yup added comments as you mentioned in the ship example

                		secondField: int
			},
		]
	},
]
```

## Generating Go Types

Once we have a schema defined, We can use it to generate a type in Go. Use the below command to generate a Go type

```
thema lineage gen gotypes -l ship.cue > go_type.go 
```

This should generate a go file has a type `ship` defined . The genearted file looks like below

```go
// This file is autogenerated. DO NOT EDIT.
//
// Generated by "thema lineage gen" from lineage defined in ship.cue

package ship

// Ship defines model for ship.
type Ship struct {
	FirstField  string `json:"firstField"`
	SecondField int    `json:"secondField"`
}
```

> If we add another schema to the sequence in `ship.cue` and again generate the Go type, Go will lose track of the first schema and will represent the second schema only.

Author: Got it

```
thema lineage gen gobindings -l ship.cue > go_bindings.go
```

The generated Go bindings will look like this (generated comments have been removed):

```go
package ship

import (
	"embed"
	"path"

	"github.com/grafana/thema"
	"github.com/grafana/thema/load"
)

//go:embed ship.cue
var themaFS embed.FS

func Lineage(rt *thema.Runtime, opts ...thema.BindOption) (thema.Lineage, error) {
	// Load a build instance from the embedded fs
	inst, err := load.InstancesWithThema(themaFS, path.Dir("ship.cue"))
	if err != nil {
		return nil, err
	}

	raw := rt.Context().BuildInstance(inst)
	return thema.BindLineage(raw, rt)
}

// type guards

var _ thema.LineageFactory = Lineage
```
We now have a single Go function our program can call, and it will load up the `ship.cue` from disk and return a [`thema.Lineage`](https://pkg.go.dev/github.com/grafana/thema#Lineage)
We can then use the Lineage function to validate our data against the schema of choice

## Validating data

Using the Thema CLI, Users can validate the schema of choice defined in `ship.cue`. Use the below command to validate the schema against an invalid JSON data 

```

> curl $URL > invalid.json

Reviewer 2: What is $URL?

Author: Placeholder for now, but added the Url for once the PR is merged
